src/Http/Http/src/StreamPipeReader.cs

```diff
@@ -0,0 +1,337 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using System.Diagnostics;
+using System.IO;
+using System.IO.Pipelines;
+using System.Runtime.CompilerServices;
+using System.Runtime.ExceptionServices;
+using System.Runtime.InteropServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.Http
+{
+    /// <summary>
+    /// Implements PipeReader using an underlying stream.
+    /// </summary>
+    public class StreamPipeReader : PipeReader
+    {
+        private readonly int _minimumSegmentSize;
+        private readonly Stream _readingStream;
+        private readonly MemoryPool<byte> _pool;
+
+        private CancellationTokenSource _internalTokenSource;
+        private bool _isCompleted;
+        private ExceptionDispatchInfo _exceptionInfo;
+
+        private BufferSegment _readHead;
+        private int _readIndex;
```

Comment:
Why don't we use the _readHead.Start instead of _readIndex?

Reply:
I thought the same, however I am running into failures when creating the ReadOnlySequence when I use Start.